File: Example/Tests/Objective-C/NyanNyanCat.m
//
//  NyanNyanCat.m
//  SuperMessagingProxyTests
//
//  Created by Aleksandr Gusev on 20/05/16.
//  Copyright © 2016 Alexander Gusev. All rights reserved.
//

#import "NyanNyanCat.h"

NS_ASSUME_NONNULL_BEGIN

@implementation NyanNyanCat

@synthesize awesomenessLevel = _awesomenessLevel;

- (instancetype)init {
    if (self = [super init]) {
        _awesomenessLevel = NSIntegerMax;
    }
    return self;
}

- (NSString *)exclamation {
    return @"Nyan! Nyan! Nyan!!!";
}

- (NSString *)says {
    return @"Nyan-nyan";
}

- (CatDescriptor)descriptor {
    CatDescriptor descriptor;
    memset(&descriptor, 0, sizeof(descriptor));
    strncpy(descriptor.name, "NyanNyanCat", sizeof("NyanNyanCat"));
    return descriptor;
}

- (NSString *)methodFromALeafClass {
    return @"I am a nyan-nyan-nyan cat!";
}

+ (NSString *)classSays {
    return @"Class nyan-nyan";
}

+ (NSInteger)classAwesomenessLevel {
    return NSIntegerMax - 1;
}

@end

NS_ASSUME_NONNULL_END

File: Example/Tests/Swift/NyanNyanCat.swift
//
//  NyanNyanCat.swift
//  SGVSuperMessagingProxy
//
//  Created by Aleksandr Gusev on 19/05/16.
//  Copyright © 2016 Alexander Gusev. All rights reserved.
//

class NyanNyanCat: NyanCat {
    override func says() -> String {
        return "Nyan-nyan"
    }
    
    override var exclamation: String {
        return "Nyan! Nyan! Nyan!!!"
    }
    
    private var _awesomenessLevel: Int = Int.max
    
    override var awesomenessLevel: Int {
        get {
            return _awesomenessLevel
        }
        set {
            _awesomenessLevel = newValue
        }
    }
    
    override class func classSays() -> String {
        return "Class Nyan-nyan"
    }
    
    override class var classAwesomenessLevel: Int {
        return Int.max
    }
}
